cpp/inc/bond/ext/grpc/wait_callback.h

```diff
@@ -0,0 +1,175 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#pragma once
+
+#include <bond/core/bonded.h>
+#include <grpc++/impl/codegen/status.h>
+
+#include <boost/optional.hpp>
+
+#include <chrono>
+#include <condition_variable>
+#include <functional>
+#include <memory>
+#include <mutex>
+
+namespace bond { namespace ext { namespace gRPC {
+
+/// @brief A callback type that can be manually waited upon.
+///
+/// The \ref wait member function can be used to wait until the callback
+/// has been called. Then, the \ref status and \ref response member
+/// functions can be called to inspect the results.
+///
+/// @example wait_callback.cpp
+template <typename TResponse>
+class wait_callback
+{
+public:
+    wait_callback() : _impl(std::make_shared<impl>()) { }
+
+    using CallbackType = std::function<void(const bond::bonded<TResponse>&, const grpc::Status&)>;
+
+    /// @brief Creates a callback that will signal this wait_callback
+    /// instance.
+    ///
+    /// If multiple callbacks are created and invoked, only the results of
+    /// the first one that is invoked will be recorded.
+    ///
+    /// It is safe for callbacks to outlive the wait_callback that created
+    /// them. When they are invoked their results will be discarded.
+    CallbackType callback()
+    {
+        // Create a weak_ptr to the implementation for the callback. If all
+        // the wait_callback instances sharing the implementation go away,
+        // their callbacks can still safely be invoked: they just won't be
+        // able to lock the weak_ptr and thus won't do anything.
+        std::weak_ptr<impl> wpImpl(_impl);
+
+        return std::function<void(const bond::bonded<TResponse>&, const grpc::Status&)>(
+            [wpImpl](const bond::bonded<TResponse>& response, const grpc::Status& status)
+            {
+                std::shared_ptr<impl> impl = wpImpl.lock();
+
+                if (impl)
+                {
+                    impl->set(response, status);
+                }
+            });
+    }
+
+    /// @brief Implicitly creates a callback.
+    operator CallbackType()
+    {
+        return callback();
+    }
+
+    /// @brief Waits for a callback created by \ref callback to be
+    /// invoked.
+    void wait() const
+    {
+        _impl->wait();
+    }
+
+    /// @brief Waits at least \p timeout for a callback created by \ref
+    /// callback to be invoked.
+    ///
+    /// @param timeout the minimum amount of time to wait.
+    ///
+    /// @return \p true if a callback was invoked. \p false if the timeout
+    /// occured.
+    template <typename Rep, typename Period>
+    bool wait(std::chrono::duration<Rep, Period> timeout) const
+    {
+        return _impl->wait(timeout);
+    }
+
+    /// @brief Gets the response.
+    ///
+    /// @warning Blocks until a callback is invoked.
+    const bond::bonded<TResponse>& response() const
```

Comment:
I don't like the name Result: there are two pieces of information here: the status code and the response.